core/src/main/scala/shapeless/generic.scala

```diff
@@ -380,82 +381,85 @@ trait CaseClassMacros extends ReprTypes {
   }
 
   def ctorsOfAux(tpe: Type, hk: Boolean): List[Type] = {
-    def collectCtors(classSym: ClassSymbol): List[ClassSymbol] = {
-      classSym.knownDirectSubclasses.toList flatMap { child0 =>
+    def collectCtors(classSym: ClassSymbol): List[ClassSymbol] =
+      classSym.knownDirectSubclasses.toList.flatMap { child0 =>
         val child = child0.asClass
         child.typeSignature // Workaround for <https://issues.scala-lang.org/browse/SI-7755>
-        if (isCaseClassLike(child) || isCaseObjectLike(child))
-          List(child)
-        else if (child.isSealed)
-          collectCtors(child)
-        else
-          abort(s"$child is not case class like or a sealed trait")
+        if (isCaseClassLike(child) || isCaseObjectLike(child)) List(child)
+        else if (child.isSealed) collectCtors(child)
+        else abort(s"$child is not case class like or a sealed trait")
       }
-    }
 
-    if(isProduct(tpe))
+    if (isProduct(tpe)) {
       List(tpe)
-    else if(isCoproduct(tpe)) {
+    } else if (isCoproduct(tpe)) {
       val basePre = prefix(tpe)
       val baseSym = classSym(tpe)
-      val baseTpe =
-        if(!hk) tpe
-        else {
-          val tc = tpe.typeConstructor
-          val paramSym = tc.typeParams.head
-          val paramTpe = paramSym.asType.toType
-          appliedType(tc, paramTpe)
-        }
+      val baseTpe = if (!hk) tpe else {
+        val tc = tpe.typeConstructor
+        appliedType(tc, tc.typeParams.head.asType.toType)
+      }
+
       val baseArgs = baseTpe.dealias.typeArgs
+      def ctorOf(sym: ClassSymbol) = {
+        import c.internal._
 
-      val ctorSyms = collectCtors(baseSym).sortBy(_.fullName)
-      val ctors =
-        ctorSyms flatMap { sym =>
-          def normalizeTermName(name: Name): TermName =
-            TermName(name.toString.stripSuffix(" "))
-
-          def substituteArgs: List[Type] = {
-            val subst = c.internal.thisType(sym).baseType(baseSym).typeArgs
-            sym.typeParams.map { param =>
-              val paramTpe = param.asType.toType
-              baseArgs(subst.indexWhere(_.typeSymbol == paramTpe.typeSymbol))
-            }
+        def substituteArgs = {
+          val subst = thisType(sym).baseType(baseSym).typeArgs
+          for (param <- sym.typeParams)
+            yield baseArgs(subst.indexWhere(_.typeSymbol == param))
+        }
+
+        // Handle anonymous subclasses.
+        val isNamed = !isAnonOrRefinement(sym)
+        if (isNamed && sym.isStatic) {
+          if (sym.isModuleClass) sym.toType
+          else appliedType(sym.toType, substituteArgs)
+        } else {
+          // Look for a path from the macro call site to the subtype.
+          val owner = sym.owner
+          val owners = ownerChain(if (isNamed) owner else owner.owner)
+          val prePaths = for (pre <- Iterator.iterate(basePre)(prefix).takeWhile(_ != NoPrefix))
+            yield (pre, owners.iterator.dropWhile(pre.baseType(_) == NoType))
```

Comment:
There are two thing going on here:
1. We try all (sub)prefixes, so that if the super class is `outer.inner.T`, we look for the subclass both in `inner` and in `outer`, in that order. Before it was only `inner`.
2. For each prefix we try to find a path to the subclass. Such a path exists if one of the classes in the owner chain is a super class of the prefix. The change here is to use `baseType` instead of symbol equality (to accomodate inheritance).